app/models/source.rb

```diff
@@ -44,63 +44,31 @@ class Source < ActiveRecord::Base
   # without setting the reimport flag. If someone really wants to turn off
   # reimporting, they should edit the source.
   def self.find_or_create_from(attrs={})
-    if attrs && attrs[:url]
-      source = Source.find_or_create_by_url(attrs[:url])
-      attrs.each_pair do |key, value|
-        if key.to_sym == :reimport
-          source.reimport = true if ! source.reimport && value
-        else
-          source.send("#{key}=", value) if source.send(key) != value
-        end
-      end
-      source.save if source.changed?
-      return source
-    else
-      return Source.new(attrs)
-    end
+    return new(attrs) unless attrs[:url]
+
+    source = find_or_create_by_url(attrs[:url])
+    source.reimport = true if attrs.delete(:reimport)
+    source.attributes = attrs
+    source.save if source.changed?
+    source
   end
 
   # Create events for this source. Returns the events created. URL must be set
   # for this source for this to work.
   def create_events!(opts={})
-    cutoff = Time.now.yesterday # All events before this date will be skipped
-    events = []
-    self.to_events(opts).each do |event|
-      if opts[:skip_old]
-        next if event.title.blank? && event.description.blank? && event.url.blank?
-        next if event.old?
-      end
-
-      # Skip invalid events that start after they end
-      next if event.end_time && event.end_time < event.start_time
-
-      # convert to local time, because time zone is simply discarded when event is saved
-      event.start_time.localtime
-      event.end_time.localtime if event.end_time
-
-      # clear duplicate_of_id field in case to_events picked up orphaned duplicate
-      # TODO clear the duplicate_of_id at the point where the object is created, not down here
-      event.duplicate_of_id = nil if event.duplicate_of_id
-      event.save!
-      if event.venue
-        event.venue.duplicate_of_id = nil if event.venue.duplicate_of_id
-        event.venue.save! if event.venue
-      end
-      events << event
-    end
-    self.save!
-    return events
+    save!
+    events = to_events(opts).select(&:valid?)
+    events.reject!(&:old?) if opts[:skip_old]
+    events.each(&:save!)
```

Comment:
It might be a little clearer to reference events explicitly for the return value, instead of relying on the iterator's return.